contracts/test/v0.8/dev/UpkeepTranscoder3_0.test.ts

```diff
@@ -0,0 +1,458 @@
+        transcoder.transcodeUpkeeps(
+          UpkeepFormat.V2,
+          UpkeepFormat.V4,
+          encodedData,
+        ),
+        'function was called with incorrect parameters',
+      )
+    })
+
+    context('when from and to versions equal', () => {
+      it('returns the data that was passed in', async () => {
+        let response = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V1,
+          UpkeepFormat.V1,
+          encodedData,
+        )
+        assert.equal(encodedData, response)
+
+        response = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V2,
+          UpkeepFormat.V2,
+          encodedData,
+        )
+        assert.equal(encodedData, response)
+
+        response = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V3,
+          UpkeepFormat.V3,
+          encodedData,
+        )
+        assert.equal(encodedData, response)
+      })
+    })
+
+    context('when from and to versions are correct', () => {
+      upkeepsV3 = [
+        [executeGas, 2 ** 32 - 1, false, target0, amountSpent, balance, 0],
+        [executeGas, 2 ** 32 - 1, false, target1, amountSpent, balance, 0],
+      ]
+
+      it('transcodes V1 upkeeps to V3 properly', async () => {
+        upkeepsV1 = [
+          [
+            balance,
+            lastKeeper0,
+            executeGas,
+            2 ** 32,
+            target0,
+            amountSpent,
+            await admin0.getAddress(),
+          ],
+          [
+            balance,
+            lastKeeper1,
+            executeGas,
+            2 ** 32,
+            target1,
+            amountSpent,
+            await admin1.getAddress(),
+          ],
+        ]
+
+        const data = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V1,
+          UpkeepFormat.V3,
+          encodeUpkeepV1(idx, upkeepsV1, ['0xabcd', '0xffff']),
+        )
+        assert.equal(
+          encodeUpkeepV3(idx, upkeepsV3, ['0xabcd', '0xffff'], admins),
+          data,
+        )
+      })
+
+      it('transcodes V2 upkeeps to V3 properly', async () => {
+        upkeepsV2 = [
+          [
+            balance,
+            lastKeeper0,
+            amountSpent,
+            await admin0.getAddress(),
+            executeGas,
+            2 ** 32 - 1,
+            target0,
+            false,
+          ],
+          [
+            balance,
+            lastKeeper1,
+            amountSpent,
+            await admin1.getAddress(),
+            executeGas,
+            2 ** 32 - 1,
+            target1,
+            false,
+          ],
+        ]
+
+        const data = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V2,
+          UpkeepFormat.V3,
+          encodeUpkeepV2(idx, upkeepsV2, ['0xabcd', '0xffff']),
+        )
+        assert.equal(
+          encodeUpkeepV3(idx, upkeepsV3, ['0xabcd', '0xffff'], admins),
+          data,
+        )
+      })
+
+      it('migrates upkeeps from one registry to another', async () => {
```

Comment:
love this test! Goes through the complete e2e flow.

Can you also add one from 1.3 to 2.0

Reply:
Done